### PR TITLE
Use configuration interfaces in FrontstageProvider

### PR DESCRIPTION
**Pull request by GerardasB**
Currently to define a `Frontstage` a number of components are used: `Frontstage`, `Zone`, `StagePanel`, `Widget`.
Elements of these components are never rendered and are only used to read the configuration data.

This PR deprecates these pseudo components in favor of configuration interfaces, allowing to define a `Frontstage` with plain JS objects:

```
// Before
class CustomFrontstageProvider extends FrontstageProvider {
  // ...

  public override get frontstage(): React.ReactElement<FrontstageProps> {
    return <Frontstage 
      id="CustomFrontstage" 
      // ...
    />;
  }
}

// After
class CustomFrontstageProvider extends FrontstageProvider {
  // ...

  public override get frontstage(): FrontstageConfig {
    return { 
      id: "CustomFrontstage",
      // ...
    };
  }
}
```

### CODE REVIEW

File: ui/appui-react/src/appui-react/content/ContentGroup.tsx [hunk 1]
```diff
@@ -44,7 +45,7 @@ export interface ContentGroupProps {
  * @public
  */
 export abstract class ContentGroupProvider {
-  abstract provideContentGroup(props: FrontstageProps): Promise<ContentGroup>;
+  abstract provideContentGroup(props: FrontstageProps | FrontstageConfig): Promise<ContentGroup>; // eslint-disable-line deprecation/deprecation
```
Comment (GerardasB) [inline ui/appui-react/src/appui-react/content/ContentGroup.tsx:48]: @raplemie this might hurt at runtime if i.e. a custom `ContentGroupProvider` is updated to use a configuration, but the `FrontstageProvider` is not (or vice versa). The TS is happy to narrow down the types though..

File: ui/appui-react/src/appui-react/frontstage/FrontstageDef.tsx [hunk 2]
```diff
@@ -108,6 +115,15 @@ export class FrontstageDef {
   public get bottomCenter(): ZoneDef | undefined { return this._bottomCenter; }
   public get bottomRight(): ZoneDef | undefined { return this._bottomRight; }
```
Comment (GerardasB) [inline ui/appui-react/src/appui-react/frontstage/FrontstageDef.tsx:116]: @raplemie we still have some properties that need to be deprecated, let's do that in a separate PR..

File: common/api/appui-react.api.md [hunk 3]
```diff
@@ -4178,6 +4218,28 @@ export class OpenMessageCenterEvent extends UiEvent<{}> {
 // @internal
 export function packNineZoneState(state: NineZoneState): SavedNineZoneState;
 
+// @beta
+export interface PanelConfig {
```
Comment (GerardasB) [inline common/api/appui-react.api.md:4222]: @raplemie, @NancyMcCallB what do you think about the naming? Should this be a `StagePanelConfig` instead for consistency? I.e. we have a `StagePanelDef`. Ideally, I'd prefer to drop the `Stage` part/prefix..

**Review by pmconne — APPROVED** [review]

File: ui/appui-react/src/appui-react/frontstage/FrontstageDef.tsx [hunk 4]
```diff
@@ -674,6 +697,39 @@ export class FrontstageDef {
...
+    if (config.applicationData !== undefined)
+      this._applicationData = config.applicationData;
+
+    this._usage = config.usage;
+    this._version = config.version || 0;
+
+    this._toolSettings = createWidgetDef(config.toolSettings, `uifw-toolSettings-widget`, WidgetType.ToolSettings);
+    this._statusBar = createWidgetDef(config.statusBar, `uifw-statusBar-widget`, WidgetType.StatusBar);
+    this._contentManipulation = createWidgetDef(config.contentManipulation, `uifw-contentManipulation-widget`, WidgetType.Tool);
+    this._viewNavigation = createWidgetDef(config.viewNavigation, `uifw-viewNavigation-widget`, WidgetType.Navigation);
```
Comment (raplemie) [inline ui/appui-react/src/appui-react/frontstage/FrontstageDef.tsx:726]: I think these should still populate the "zone" version of this, otherwise the fronstage is branching and that might break code that still use the previous stuff, I dont know to which extent it would cause a problem. (Could a package provided frontstage be interracting with an application that would not be updated and still look at "topLeft" instead of "contentManipulation" ?

File: ui/appui-react/src/appui-react/frontstage/FrontstageDef.tsx [hunk 5]
```diff
@@ -1086,3 +1142,37 @@ export class FrontstageDef {
...
   }
 }
+
+function createWidgetDef(config: WidgetConfig | undefined, stableId: string, type: WidgetType): WidgetDef | undefined {
+  if (!config)
+    return undefined;
+
+  const stableConfig = getStableWidgetProps(config, stableId);
+  const widgetDef = new WidgetDef(stableConfig);
+  widgetDef.initializeFromConfig(stableConfig, type);
```
Comment (aruniverse) [inline ui/appui-react/src/appui-react/frontstage/FrontstageDef.tsx:1152]: does this seem unecessary? why not just pass the type to the constructor and encapsulate all of this?
same goes for the panel

File: ui/appui-react/src/appui-react/frontstage/FrontstageConfig.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  /** Id for the Frontstage */
+  readonly id: string;
+  /** Tool that is started once the Frontstage is activated */
+  readonly defaultTool: ToolItemDef;
+  /** The Content Group providing the Content Views */
+  readonly contentGroup: ContentGroup | ContentGroupProvider;
+  /** Id of the Content View to be activated initially */
+  readonly defaultContentId?: string;
+  /** Any application data to attach to this Frontstage. */
+  readonly applicationData?: any;
```
Comment (aruniverse) [inline ui/appui-react/src/appui-react/frontstage/FrontstageConfig.ts:28]: do we want to bring this fwd to the new api?

File: ui/appui-react/src/appui-react/frontstage/FrontstageConfig.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+  /** Usage type for this Frontstage. */
+  readonly usage?: string;
+  /** Frontstage version. Used to force saved layout reinitialization after changes to frontstage.
+   * @note This value should be increased when changes are made to Frontstage.
+   * Increasing the value will make sure to reinitialize App layout instead of restoring to old layout.
+   * Version increase is required when widgets are added/removed.
+   */
+  readonly version?: number;
```
Comment (aruniverse) [inline ui/appui-react/src/appui-react/frontstage/FrontstageConfig.ts:36]: this should be required; when its optional most of us never set it; and will never think to update this

File: ui/appui-react/src/appui-react/content/ContentGroup.tsx (hunk 1 above)
Comment (raplemie) [reply ui/appui-react/src/appui-react/content/ContentGroup.tsx:48]: Yes, we cant have that like it given that it is possible to mix and match providers in frontstages that we dont own.

I think we'd need a different method name to get away with clashes, but that will still require us to be able to convert from a Props to a Config so we can pass it whatever it expects, which is not a simple thing.

```suggestion
  /**
   * Return the contentGroup based on the `FrontstageProps` received.
   * @deprecated Implement using `contentGroup(FrontstageConfig)` instead, no need to provide this method if the other is provided. At least one should be provided in order for the provider to do anything.
   */
  abstract provideContentGroup?(props: FrontstageProps): Promise<ContentGroup>; // eslint-disable-line deprecation/deprecation
  
  /**
   * Return the contentGroup based on the `FrontstageConfig`.
   * @beta This method will be required in upcoming version, this method will be prioritized if it exists over `provideContentGroup`. Only one of these 2 methods will be called and at least one need to be implemented in order for the provider to do anything.)
   */
  abstract contentGroup?(config: FrontstageConfig): Promise<ContentGroup>;
```

File: ui/appui-react/src/appui-react/frontstage/FrontstageConfig.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  /** if isIModelIndependent then frontstage is independent from any iModel. */
+  readonly isIModelIndependent?: boolean;
```
Comment (aruniverse) [inline ui/appui-react/src/appui-react/frontstage/FrontstageConfig.ts:38]: i wasnt aware we had this; what is the use case for this?

File: ui/appui-react/src/appui-react/widgets/WidgetConfig.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+/** @packageDocumentation
+ * @module Widget
+ */
+
+import { WidgetProps } from "./WidgetProps";
+
+/** Configuration from which a widget is created.
+ * @beta
+ */
+export type WidgetConfig = WidgetProps; // eslint-disable-line deprecation/deprecation
```
Comment (aruniverse) [inline ui/appui-react/src/appui-react/widgets/WidgetConfig.ts:14]: i dont get the purpose of this? is this just an alias/rename we want to do for 4.0 to keep the Config suffix name consistency ?

**Review by raplemie — CHANGES_REQUESTED** [review]
Seems that if we go with as it is currently, we'll get into trouble, dont have a strong suggestion to improve though, only ideas.

File: ui/appui-react/src/appui-react/widgets/WidgetConfig.ts (hunk 9 above)
Comment (raplemie) [reply ui/appui-react/src/appui-react/widgets/WidgetConfig.ts:14]: Yes, given that this is a UI package, props are typically "parameters of a react component", which is no longer the case, that's why we changed the name.

**Comment by aruniverse** [comment]
my sincere apologies; didnt realize when i was reviewing on github.dev all the comments would come in separately as individual reviews and not all together

File: ui/appui-react/src/appui-react/frontstage/FrontstageConfig.ts [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+ */
+
+import { CommonProps } from "@itwin/core-react";
+import { ContentGroup, ContentGroupProvider } from "../content/ContentGroup";
+import { ToolItemDef } from "../shared/ToolItemDef";
+import { PanelConfig } from "../stagepanels/PanelConfig";
+import { WidgetConfig } from "../widgets/WidgetConfig";
+
+/** Configuration from which a frontstage is created.
+ * @public
```
[the remark this answers is not in the document]
Comment (GerardasB) [reply ui/appui-react/src/appui-react/frontstage/FrontstageConfig.ts:16]: Can't ATM, since `FrontstageProvider.frontstage` is marked as **public**..
Returning that union type from the `FrontstageProvider.frontstage` is also breaking, I think we should also add an optional getter that would return a new config instead. Any suggestions on the naming?

File: ui/appui-react/src/appui-react/frontstage/FrontstageDef.tsx (hunk 4 above)
Comment (GerardasB) [reply ui/appui-react/src/appui-react/frontstage/FrontstageDef.tsx:726]: You are right, that conversion step makes sense, especially when gradually reacting to the changes in 3.x. Will work on that.

File: ui/appui-react/src/appui-react/frontstage/FrontstageConfig.ts (hunk 8 above)
Comment (GerardasB) [reply ui/appui-react/src/appui-react/frontstage/FrontstageConfig.ts:38]: `version` is to invalidate the saved/chached frontstage version.
Not sure what the use-case is for `isIModelIndependent` (added by https://github.com/iTwin/itwinjs-core/pull/2423), @NancyMcCallB?

File: ui/appui-react/src/appui-react/frontstage/FrontstageConfig.ts (hunk 7 above)
Comment (GerardasB) [reply ui/appui-react/src/appui-react/frontstage/FrontstageConfig.ts:36]: Makes sense to me. I think we should make `WidgetConfig["id"]` required as well..

File: common/api/appui-react.api.md (hunk 3 above)
Comment (raplemie) [reply common/api/appui-react.api.md:4222]: I wonder if panel wouldnt be too generic, maybe `SidePanelConfig` ? (But, it may not be necessary to change at all)

File: ui/appui-react/src/appui-react/frontstage/FrontstageDef.tsx [hunk 11]
```diff
@@ -674,6 +697,39 @@ export class FrontstageDef {
...
+      this._defaultContentId = config.defaultContentId;
+
+    if (config.contentGroup instanceof ContentGroupProvider) {
+      this._contentGroupProvider = config.contentGroup;
+    } else {
+      this._contentGroup = config.contentGroup;
+    }
+
+    if (config.applicationData !== undefined)
+      this._applicationData = config.applicationData;
```
[the remark this answers is not in the document]
Comment (GerardasB) [reply ui/appui-react/src/appui-react/frontstage/FrontstageDef.tsx]: Removed the check.

File: ui/appui-react/src/appui-react/frontstage/FrontstageDef.tsx [hunk 12]
```diff
@@ -65,6 +68,10 @@ export class FrontstageDef {
   private _applicationData?: any;
   private _usage?: string;
   private _version: number = 0;
+  private _toolSettings?: WidgetDef;
+  private _statusBar?: WidgetDef;
+  private _contentManipulation?: WidgetDef;
+  private _viewNavigation?: WidgetDef;
```
[the remark this answers is not in the document]
Comment (GerardasB) [reply ui/appui-react/src/appui-react/frontstage/FrontstageDef.tsx]: Refactored so the private members, initialization and all the data structures where we store the layout/widgets remain as is - based on `FrontstageProps`. There are a couple of internal helpers to transform between `FrontstageConfig` and `FrontstageProps`.
I will add additional tests to verify that all the properties are set correctly when using the helpers.

File: ui/appui-react/src/appui-react/frontstage/FrontstageDef.tsx (hunk 4 above)
Comment (GerardasB) [reply ui/appui-react/src/appui-react/frontstage/FrontstageDef.tsx]: Refactored as suggested.

File: ui/appui-react/src/appui-react/frontstage/FrontstageConfig.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (GerardasB) [reply ui/appui-react/src/appui-react/frontstage/FrontstageConfig.ts]: Well, technically it's a public class, so we have to assume that someone might use it outside of our code base.
Until we decide on the naming I have added a separate method to get the config: `public frontstageConfig?(): FrontstageConfig;`.

File: ui/appui-react/src/appui-react/frontstage/FrontstageDef.tsx [hunk 13]
```diff
@@ -272,8 +291,14 @@ export class FrontstageDef {
   public async onActivated() {
     this.updateWidgetDefs();
 
-    if (this._contentGroupProvider && this._initialProps) {
-      this._contentGroup = await this._contentGroupProvider.provideContentGroup(this._initialProps);
+    const provider = this._contentGroupProvider;
+    if (provider && this._initialProps) {
+      if (provider.contentGroup) {
+        const config = toFrontstageConfig(this._initialProps);
```
Comment (raplemie) [inline ui/appui-react/src/appui-react/frontstage/FrontstageDef.tsx:297]: Since the frontstage can be activated more than once, should we save the config, so we can reuse it on subsequent calls ?

File: ui/appui-react/src/appui-react/content/ContentGroup.tsx (hunk 1 above)
Comment (raplemie) [reply ui/appui-react/src/appui-react/content/ContentGroup.tsx]: Seems all good

File: ui/appui-react/src/appui-react/frontstage/FrontstageConfig.ts (hunk 7 above)
Comment (GerardasB) [reply ui/appui-react/src/appui-react/frontstage/FrontstageConfig.ts]: FrontstageConfig["version"] and WidgetConfig["id"] properties are changed to required.

File: ui/appui-react/src/appui-react/frontstage/FrontstageConfig.ts (hunk 6 above)
Comment (GerardasB) [reply ui/appui-react/src/appui-react/frontstage/FrontstageConfig.ts]: For now I have removed these properties:
**`FrontstageConfig`**: `defaultContentId`, `applicationData`, `isIModelIndependent`, `defaultTool`.
**`PanelConfig`**: `applicationData`, `header`.

File: ui/appui-react/src/appui-react/shared/ActionButtonItemDef.tsx [hunk 14]
```diff
@@ -90,3 +89,5 @@ export abstract class ActionButtonItemDef extends ItemDefBase {
     );
   }
 }
+
+import { ActionItemButton } from "../toolbar/ActionItemButton";
```
Comment (aruniverse) [inline ui/appui-react/src/appui-react/shared/ActionButtonItemDef.tsx:93]: should be able to revert this change

File: ui/appui-react/src/appui-react/frontstage/FrontstageDef.tsx (hunk 13 above)
Comment (GerardasB) [reply ui/appui-react/src/appui-react/frontstage/FrontstageDef.tsx:298]: It costs nothing to convert between Props and Config, let's keep this branch free.

File: ui/appui-react/src/appui-react/shared/ActionButtonItemDef.tsx (hunk 14 above)
Comment (GerardasB) [reply ui/appui-react/src/appui-react/shared/ActionButtonItemDef.tsx:93]: Reverting this leads to a circular dependency. Is there a problem w/ this?

Comment (aruniverse) [reply ui/appui-react/src/appui-react/shared/ActionButtonItemDef.tsx:93]: gotcha, i wasnt sure if this was by accident or circular dep. long term we probably want to fix this

File: common/api/appui-react.api.md (hunk 3 above)
Comment (GerardasB) [reply common/api/appui-react.api.md]: I have renamed `PanelConfig` to `StagePanelConfig` for consistency, since we have a bunch of panel related types prefixed with `StagePanel`, i.e. `StagePanelLocation`, `StagePanelSection`...

File: docs/changehistory/NextVersion.md [hunk 15]
```diff
@@ -183,7 +183,15 @@ All non-internal components are deprecated with their corresponding replacements
...
-`NineZoneChangeHandler`, `StagePanelChangeHandler`, `WidgetStateFunc`, `ZoneDefProvider`.
+`NineZoneChangeHandler`, `StagePanelChangeHandler`, `WidgetStateFunc`, `ZoneDefProvider`, `Zone`, `ZoneDef`.
+
+Pseudo components used by the [FrontstageProvider]($appui-react) are deprecated and replaced by corresponding configuration interfaces:
+
+| Component    | Replacement                      |
+| ------------ | -------------------------------- |
+| `Frontstage` | [FrontstageConfig]($appui-react) |
+| `Widget`     | [WidgetConfig]($appui-react)     |
+| `StagePanel` | [StagePanelConfig]($appui-react) |
```
Comment (aruniverse) [inline docs/changehistory/NextVersion.md:194]: arent `StagePanels` a `UI 1.0` concept? Completely replaced by widgets?
If so, do we eve need a replacement? Just mark as deprecated and completely remove in 4.0?

Comment (GerardasB) [reply docs/changehistory/NextVersion.md:194]: We still have top/bottom/left/right panels in UI2.0 (where widgets can be docked). I was wondering if it would make sense to rename these to `PanelDef`, `PanelConfig` instead of using existing `StagePanelDef`..

**Review by NancyMcCallB — APPROVED** [review]

File: ui/appui-react/src/appui-react/frontstage/FrontstageConfig.ts (hunk 8 above)
Comment (aruniverse) [reply ui/appui-react/src/appui-react/frontstage/FrontstageConfig.ts]: imo these should never have been inside a stage; so good thing we're removing it

**Review by raplemie — APPROVED** [review]

File: ui/appui-react/src/appui-react/shared/ActionButtonItemDef.tsx (hunk 14 above)
Comment (aruniverse) [reply ui/appui-react/src/appui-react/shared/ActionButtonItemDef.tsx:93]: i _think_ this change here is breaking the viewer and probably anyone using appui-react with the 3.5 rc
investigating...
image

Comment (aruniverse) [reply ui/appui-react/src/appui-react/shared/ActionButtonItemDef.tsx:93]: able to repro on master w/ ui-test-app.
unable to repro w/ `acb7d2654d9533eb3e898c628951f43b15ef07ff` (commit right before this pr went through)
